lxd/migrate_container.go

```diff
@@ -684,28 +731,47 @@ func (s *migrationSourceWs) Do(state *state.State, migrateOp *operations.Operati
 			}
 		}
 
-		// We do the transger serially right now, but there's really no reason for us to;
+		// We do the transfer serially right now, but there's really no reason for us to;
 		// since we have separate websockets, we can do it in parallel if we wanted to.
 		// However assuming we're network bound, there's really no reason to do these in.
 		// parallel. In the future when we're using p.haul's protocol, it will make sense
 		// to do these in parallel.
 		ctName, _, _ := shared.InstanceGetParentAndSnapshotName(s.instance.Name())
-		state := s.instance.DaemonState()
-		err = rsync.Send(ctName, shared.AddSlash(checkpointDir), &shared.WebsocketIO{Conn: s.criuConn}, nil, rsyncFeatures, bwlimit, state.OS.ExecPath)
+		err = rsync.Send(ctName, shared.AddSlash(checkpointDir), &shared.WebsocketIO{Conn: s.criuConn}, nil, rsyncFeatures, rsyncBwlimit, state.OS.ExecPath)
 		if err != nil {
 			return abort(err)
 		}
 	}
 
-	if pool == nil {
-		if s.live || (respHeader.Criu != nil && *respHeader.Criu == migration.CRIUType_NONE) {
-			err = legacyDriver.SendAfterCheckpoint(s.fsConn, bwlimit)
+	// If s.live is true or Criu is set to CRIUTYPE_NONE rather than nil, it indicates
+	// that the source container is running and that we should do a two stage transfer
+	// to minimize downtime.
+	if s.live || (respHeader.Criu != nil && *respHeader.Criu == migration.CRIUType_NONE) {
+		if pool != nil {
+			volSourceArgs := migration.VolumeSourceArgs{
+				Name:          s.instance.Name(),
+				MigrationType: migrationType,
+				TrackProgress: true,
+				FinalSync:     true,
+			}
+
+			err = pool.MigrateInstance(s.instance, &shared.WebsocketIO{Conn: s.fsConn}, volSourceArgs, migrateOp)
```

Comment:
@stgraber wanted to draw your attention to this. Rather than have two different functions that implement the equivalent of `SendWhileRunning` and `SendAfterCheckpoint` i've called the same `pool.MigrateInstance` function at both phases, but instead pass a `FinalSync: true` argument in the volSourceArgs during the `SendAfterCheckpoint` stage indicating that this is the final sync and allowing the driver to change behaviour if needed. I've also ensured that backendLXD will kick up an error if you try and sync snapshots during the final sync.

Reply:
@stgraber also, as per https://github.com/lxc/lxd/pull/6540#discussion_r353834959 this function will be called whilst the container is still running.